src/Framework/TestRunner.php

```diff
@@ -249,6 +250,174 @@ public function run(TestCase $test): void
+        while ($bytes_read < $bytes_total)
+        {
+            $read = @socket_read($socket, $bytes_total - $bytes_read);
+            if ($read === false)
+            {
+                throw new \RuntimeException('socket_receive error: ' . socket_strerror(socket_last_error()));
+            }
+
+            // blank socket_read means done
+            if ($read == '')
+            {
+                break;
+            }
+
+            $bytes_read += strlen($read);
+            $socket_message .= $read;
+
+            if (!$have_header && $bytes_read >= self::SOCKET_HEADER_SIZE)
+            {
+                $have_header = true;
+                list($bytes_total) = array_values(unpack('N', $socket_message));
+                $bytes_read = 0;
+                $socket_message = '';
+            }
+        }
+
+        return @unserialize($socket_message);
+    }
+
+    private function socket_send($socket, $message)
+    {
+        $serialized_message = @serialize($message);
+        if ($serialized_message == false)
+        {
+            throw new \RuntimeException('socket_send failed to serialize message');
+        }
+
+        $header = pack('N', strlen($serialized_message));
+        $data = $header . $serialized_message;
+        $bytes_left = strlen($data);
+        while ($bytes_left > 0)
+        {
+            $bytes_sent = @socket_write($socket, $data);
+            if ($bytes_sent === false)
+            {
+                throw new \RuntimeException('socket_send failed to write to socket');
+            }
+
+            $bytes_left -= $bytes_sent;
+            $data = substr($data, $bytes_sent);
+        }
+    }
+
+    private function runInFork(TestCase $test): void
+    {
+        list($socket_child, $socket_parent) = $this->ipc_init();
+
+        $pid = pcntl_fork();
+
+        if ($pid === -1 ) {
+            throw new \Exception('could not fork');
+        } else if ($pid) {
+            // we are the parent
+
+            socket_close($socket_parent);
+
+            // read child stdout, stderr
+            $result = $this->socket_receive($socket_child);
+
+            $stderr = '';
+            $stdout = '';
+            if (is_array($result) && array_key_exists('error', $result)) {
+                $stderr = $result['error'];
+            } else {
+                $stdout = $result;
+            }
+
+            $php = AbstractPhpProcess::factory();
+            $php->processChildResult($test, $stdout, $stderr);
+
+        } else {
+            // we are the child
+
+            socket_close($socket_child);
+
+            $offset                  = hrtime();
+            $dispatcher = Event\Facade::instance()->initForIsolation(
+                \PHPUnit\Event\Telemetry\HRTime::fromSecondsAndNanoseconds(
+                    $offset[0],
+                    $offset[1]
+                )
+            );
+
+            $test->setInIsolation(true);
+            try {
+                $test->run();
```

Comment:
I think I need to check `$runEntireClass` at this point and run only a single method. not sure how yet.
(in a similar way it is done in  `runInWorkerProcess()`)